src/tbtcv2-rewards/rewards.ts

```diff
@@ -0,0 +1,726 @@
+// rewards interval dates.
+const offset = Math.floor(Date.now() / 1000) - endRewardsTimestamp;
+
+export async function calculateRewards() {
+  if (Date.now() / 1000 < endRewardsTimestamp) {
+    console.log("End time interval must be in the past");
+    return "End time interval must be in the past";
+  }
+
+  const provider = new ethers.providers.EtherscanProvider(
+    network,
+    process.env.ETHERSCAN_TOKEN
+  );
+
+  const monthlyRate = (APR / 12) * HUNDRED; // monthly periodic rate ARP / 12 month
+  const currentBlockNumber = await provider.getBlockNumber();
+  const rewardsInterval = endRewardsTimestamp - startRewardsTimestamp;
+
+  // Query for bootstrap data that has peer instances grouped by operators
+  const queryBootstrapData = `${prometheusAPI}/query_range`;
+  const paramsBootstrapData = {
+    query: `sum by(chain_address)({job='${prometheusJob}'})`,
+    start: startRewardsTimestamp,
+    end: endRewardsTimestamp,
+    step: QUERY_STEP,
+  };
+
+  const bootstrapData = (
+    await queryPrometheus(queryBootstrapData, paramsBootstrapData)
+  ).data.result;
+
+  const operatorsData = new Array();
+  const rewardsData = new Array();
+
+  const randomBeacon = new Contract(
+    RandomBeaconAddress,
+    JSON.stringify(RandomBeaconABI),
+    provider
+  );
+
+  const tokenStaking = new Contract(
+    TokenStakingAddress,
+    JSON.stringify(TokenStakingABI),
+    provider
+  );
+
+  const walletRegistry = new Contract(
+    WalletRegistryAddress,
+    JSON.stringify(WalletRegistryABI),
+    provider
+  );
+
+  console.log("Fetching AuthorizationIncreased events in rewards interval...");
+  const intervalAuthorizationIncreasedEvents = await tokenStaking.queryFilter(
+    "AuthorizationIncreased",
+    startRewardsBlock,
+    endRewardsBlock
+  );
+
+  console.log("Fetching AuthorizationDecreased events in rewards interval...");
+  const intervalAuthorizationDecreasedEvents = await tokenStaking.queryFilter(
+    "AuthorizationDecreaseApproved",
+    startRewardsBlock,
+    endRewardsBlock
+  );
+
+  console.log(
+    "Fetching AuthorizationIncreased events after rewards interval..."
+  );
+  const postIntervalAuthorizationIncreasedEvents = await tokenStaking.queryFilter(
+    "AuthorizationIncreased",
+    endRewardsBlock,
+    currentBlockNumber
+  );
+
+  console.log(
+    "Fetching AuthorizationDecreased events after rewards interval..."
+  );
+  const postIntervalAuthorizationDecreasedEvents = await tokenStaking.queryFilter(
+    "AuthorizationDecreaseApproved",
+    endRewardsBlock,
+    currentBlockNumber
+  );
+
+  for (let i = 0; i < bootstrapData.length; i++) {
+    const operatorAddress = bootstrapData[i].metric.chain_address;
+    let authorizations = new Map<string, BigNumber>(); // application: value
+    let requirements = new Map<string, boolean>(); // factor: true | false
+    let instancesData = new Map<string, Map<string, string | number>>();
+    let operatorData: any = {};
+    let rewardData: any = {};
```

Comment:
```suggestion

```

Reply:
Sure, part of [c78a6b7](https://github.com/threshold-network/merkle-distribution/pull/35/commits/c78a6b730f5fc7045bf5fbbd34b65be39292a8f4)